src/library_pthread.js

```diff
@@ -37,33 +33,7 @@ var LibraryPThread = {
       }
 #endif
     },
-    initRuntime: function() {
-#if USE_ASAN || USE_LSAN
-      // When sanitizers are enabled, malloc is normally instrumented to call
-      // sanitizer code that checks some things about pthreads. As we are just
-      // setting up the main thread here, and are not ready for such calls,
-      // call malloc directly.
-      withBuiltinMalloc(function () {
-#endif
-
-      var tb = _malloc({{{ C_STRUCTS.pthread.__size__ }}});
-
-      for (var i = 0; i < {{{ C_STRUCTS.pthread.__size__ }}}/4; ++i) HEAPU32[tb/4+i] = 0;
-
-      // The pthread struct has a field that points to itself - this is used as
-        // a magic ID to detect whether the pthread_t structure is 'alive'.
-      {{{ makeSetValue('tb', C_STRUCTS.pthread.self, 'tb', 'i32') }}};
-
-      // pthread struct robust_list head should point to itself.
-      var headPtr = tb + {{{ C_STRUCTS.pthread.robust_list }}};
-      {{{ makeSetValue('headPtr', 0, 'headPtr', 'i32') }}};
-
-      // Allocate memory for thread-local storage.
-      var tlsMemory = _malloc({{{ cDefine('PTHREAD_KEYS_MAX') * 4 }}});
-      for (var i = 0; i < {{{ cDefine('PTHREAD_KEYS_MAX') }}}; ++i) HEAPU32[tlsMemory/4+i] = 0;
-      Atomics.store(HEAPU32, (tb + {{{ C_STRUCTS.pthread.tsd }}} ) >> 2, tlsMemory); // Init thread-local-storage memory array.
-      Atomics.store(HEAPU32, (tb + {{{ C_STRUCTS.pthread.tid }}} ) >> 2, tb); // Main thread ID.
```

Comment:
I understand where the code above for tb.self etc. has moved to (C), but what now handles the TLS init?

Reply:
Indeed I probably should have explained that.   musl uses statically allocated `tsd` for the main thread.

It gets assigned on first use here: https://github.com/emscripten-core/emscripten/blob/b752bd6838124f76083c2a37529e488fd0de3d73/system/lib/libc/musl/src/thread/pthread_key_create.c#L24

Reply:
So we can avoid malloc completely!

Reply:
Interesting, thanks... and nice to avoid malloc!

So this is for `pthread_key*` - what about C/C++ level TLS? Is that handled in another way?

Reply:
Yes, C/C++ TLS variables are handled separately to the dynamically created TLS keys in `pthread_key_create`